barretenberg/cpp/src/barretenberg/translator_vm/README.md

```diff
@@ -0,0 +1,1331 @@
+|                                                |                                                          |                                                           |
+
+The tail microlimbs (shown in yellow) enforce tight range constraints by ensuring top limbs use exactly the required number of bits (explained in the Decomposition Relation section of [RELATIONS.md](RELATIONS.md)).
+
+**Column reuse optimization:** Some columns are reassigned in odd rows to hold tail microlimbs for limbs that don't need all 5 microlimbs. For example, limb $P_{x, 3}$ is only 50 bits (= 3×14 + 8), requiring only 4 microlimbs. The 5th microlimb column `P_X_HIGH_LIMBS_RANGE_CONSTRAINT_4` at odd rows is therefore reassigned to hold the tail microlimb for $P_{x,3}$ (and carry values $c^{\text{lo}}[4]$, $c^{\text{hi}}[4]$, etc.).
+
+### Virtual Columns
+
+Some columns are "virtual" and not explicitly stored in the witness trace. Instead, they are computed on-the-fly during relation evaluation using existing columns. These include:
+
+- Interleaved columns for range constraint microlimbs (computed from the physical microlimb columns)
+- Sorted (ordered) columns for range constraint microlimbs (computed by sorting the physical microlimb columns)
+
+### Lagrange Polynomials (Precomputed)
+
+The Translator circuit uses **ZERO selector polynomials** (`NUM_SELECTORS = 0`).
+
+Instead, the circuit uses **Lagrange polynomials** to control which constraints are active:
+
+| Polynomial                     | Description                     | Active Rows                           |
+| ------------------------------ | ------------------------------- | ------------------------------------- |
+| `lagrange_even_in_minicircuit` | Even indices in mini-circuit    | $i \in \{0, 2, 4, ..., 8190\}$ (mini) |
+| `lagrange_odd_in_minicircuit`  | Odd indices in mini-circuit     | $i \in \{1, 3, 5, ..., 8191\}$ (mini) |
+| `lagrange_first`               | First row                       | $i = 0$                               |
+| `lagrange_last_in_minicircuit` | Last row in mini-circuit        | $i = 8191$ (mini)                     |
+| `lagrange_result_row`          | Row containing final result     | Specific row in trace                 |
+| `lagrange_masking`             | Masking rows for zero-knowledge | Last few rows                         |
+| `lagrange_mini_masking`        | Masking within mini-circuit     | Last rows of mini-circuit             |
+
+The circuit's regularity (2-row cycles, uniform structure) allows using Lagrange polynomials, which are more efficient than custom selectors.
+
+---
+
+## Interleaving: The Key Optimization
+
+The Translator must range-constrain approximately 64 different microlimb sets using permutation argument. The permutation argument's degree equals $1 + \text{NUM\_COLS}$, where NUM_COLS is the number of columns being permuted:
+
+$$
+z_{\textsf{perm}}[i+1] \cdot \prod_{j=1}^{\textsf{NUM\_COLS}} (\textsf{ordered}[j] + \beta + \gamma) =
```

Comment:
I think you either need to include the lagrange in the formula or remove the beta. Before ZK there was not beta